### PR TITLE
[Merged by Bors] - feat: add definitions of degrees of AddMonoidAlgebras

### PR DESCRIPTION
**Pull request by adomani**
This PR defines

1. the `supDegree` of an `AddMonoidAlgebra`, graded by a `SemilatticeSup`;
1. the `infDegree` of an `AddMonoidAlgebra`, graded by a `SemilatticeInf`;
1. ~~the `maxDegree` of an `AddMonoidAlgebra`, graded by a `LinearOrder` with a bottom element;~~
1. ~~the `minDegree` of an `AddMonoidAlgebra`, graded by a `LinearOrder` with a top element.~~

It also proves that under addition and multiplication in the `AddMonoidAlgebra`, each degree behaves as expected.

The lemmas are there simply to give a sample usage of the new definition.

The [corresponding `mathlib3` PR](https://github.com/leanprover-community/mathlib/pull/15296) was `not-too-late` and closed unmerged.

---

Please, let me know which of these definitions seems more useful, or whether you think that different assumptions would be better.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/MonoidAlgebra/Degree.lean [hunk 1]
```diff
@@ -172,4 +210,129 @@ end CommutativeLemmas
...
+theorem maxDegree_mul_le (f g : AddMonoidAlgebra R A) :
+    (f * g).maxDegree D ≤ f.maxDegree D + g.maxDegree D :=
+  sup_support_mul_le (fun {_ _} => (AddMonoidHom.map_add D _ _).le) f g
+#align add_monoid_algebra.max_degree_mul_le AddMonoidAlgebra.maxDegree_mul_le
+
+end MaxDegree
+
+section MinDegree
+
+variable [AddZeroClass A] [LinearOrderedAddCommMonoid T] [OrderTop T] (D : A →+ T)
```
Comment (eric-wieser) [inline Mathlib/Algebra/MonoidAlgebra/Degree.lean:311]: Is there a reason not to generalize this to
```suggestion
variable [AddZeroClass A] [LinearOrderedAddCommMonoid T] [OrderTop T] (D : A → T)
```

Comment (eric-wieser) [reply Mathlib/Algebra/MonoidAlgebra/Degree.lean:311]: Ah, I found https://github.com/leanprover-community/mathlib/pull/15296#issuecomment-1185746141:

> I'm not against generality; in fact, I just found one use case in mathlib where `B` is not a linear_order: [mv_polynomial.degrees](https://leanprover-community.github.io/mathlib_docs/data/mv_polynomial/variables.html#mv_polynomial.degrees), so I'm in favor of keeping the smilattice_inf versions (`D` is still a homomorphism though.)
> 
> I do think the current lemmas would be more convenient to use if specialized to the linear_order case. If they're gonna be used in 3 situations that all have `B` being linear_order then we should probably do the specialization here instead at each of the 3 situations. Maybe we should reserve `max_degree` and `min_degree` for the linear_order case, and rename them to `sup_degree` and `inf_degree` in the semilattice_sup/inf case. The linear_order version would be stated using `[linear_ordered_add_comm_monoid B]` and `D : A →+ B`, which would be more convenient to use, as the D0, Dm hypotheses cannot be supplied by typeclass inference.
> 
> There are also many lemmas that hold for linear_order but not general semilattice_inf; [polynomial.degree_add_eq_left_of_degree_lt](https://leanprover-community.github.io/mathlib_docs/data/polynomial/degree/definitions.html#polynomial.degree_add_eq_left_of_degree_lt) (consider p = xy^2-yx^2 and q = yx^2), and ~[polynomial.degree_mul'](https://leanprover-community.github.io/mathlib_docs/data/polynomial/degree/definitions.html#polynomial.degree_mul') (leading_coeff only really makes sense in linear_order case)~ (sorry, can't really state it this way unless D is injective; would need to use [graded_ring.proj](https://leanprover-community.github.io/mathlib_docs/ring_theory/graded_algebra/basic.html#graded_ring.proj)/decompose).

Comment (adomani) [reply Mathlib/Algebra/MonoidAlgebra/Degree.lean:311]: I am in favour of switching back to `SemilatticeInf`, but am not opposed to leaving it as is, given the controversial nature of the assumption!

In other situations, I remember regretting not having made a possible generalisation.  I think that I never regretted generalising too much!

Comment (jcommelin) [reply Mathlib/Algebra/MonoidAlgebra/Degree.lean:311]: Ok, let's switch back to the general version.

**Comment by jcommelin** [comment]
LGTM. Please merge after processing the comment by Eric.

bors d+

**Comment by adomani** [comment]
Johan, thank you very much for the review!

I am going to wait to merge, to give @eric-wieser and @alreadydone a chance to give comments.

The suggestion that Eric gave required some extra tweaking.  The more general definition of `max/minDegree` absorbed the older notions of `sup/infDegree`, so I removed the less general version.  That version is still available via, e.g., `f.supDegree ((↑) : A → WithBot A)`.

I went for a conservative approach: introduce only the most general definition that was proposed and not the more specialized one.  I think that allowing the degree to take place in an order that is not necessarily the grading type of the AddMonoidAlgebra will be useful: for instance, it may be useful to work with the `Nat`-grading of `MvPolynomials` and not be stuck with working necessarily with the full `sigma --> Nat`.

**Comment by alreadydone** [comment]
Hi @adomani, I've been working on the fundamental theorem in the past 3 days, which is now sorry free (so @negiizhao's Lindemann-Weierstrass proof only needs your [is_domain add_monoid_algebra](https://github.com/leanprover-community/mathlib/pull/16157) to work.

To complete the proof I developed substantial API for `maxDegree` taken values in `Lex (Finsupp σ ℕ)` which is a linear order when `σ` is given one; note the absence of `WithBot`, so it's more like `natDegree` rather than `Polynomial.degree`. Even though it's a linear order, I've defined it using `Finset.sup` like you did in this PR., which is more convenient than `Finset.max` (values in `WithBot`) and `Finset.max'` (requires `Nonempty` and therefore `if then else` in definition). If you have further work in this direction, we should discuss and collaborate on a PR introducing the API.

**Comment by adomani** [comment]
Dear @alreadydone, I am very happy that "`degree`s" related to the ones in this PR are useful!  As you saw, I ended up only merging one.

I also think that the specialized ones are great as local definitions, as in the case of your `maxDegree`.  Also, I view the difference between `degree` and `natDegree` as an inconvenience of formalization.  There should really be only one notion and enough automation to deal with `WithBot` invisibly in the background.  I know that this is not the case right now, but I find that the duplication of all the lemmas (or snaking around the available ones) not really ideal!

Finally, I was not planning to port the no-zero divisor instances soon: I am currently using my lean time to learn some meta-programming.  I would be very happy to see it in mathlib4, but honestly, it is not one of my priorities right now.

Sorry if I am making your progress harder!  Feel free to press my, in case you really want me to port the [`is_domain add_monoid_algebra](https://github.com/leanprover-community/mathlib/pull/16157) PR -- I'll try to be as responsive as I can! :smile:

**Pull request closed without merging**